url5997ccb760_JANNLab_JANNLab_tgz-pJ8-de_jannlab_examples_recurrent_SequentialXorExampleJ8/src/de/jannlab/math/Histogram.java

```diff
@@ -0,0 +1,373 @@
+     * Accumulates the given value in the current histogram with
+     * an additionally given weight.
+     * <br></br>
+     * @param value Value within the given interval.
+     * @param weight The weight of the value.
+     */
+    public void note(final double value, final double weight) {
+        final int idx  =  quantize(value);
+        if ((idx < 0) || (idx >= this.quanta)) return;
+        this.accu[idx] += weight;
+    }
+    
+    /**
+     * Resets the current histogram. All accumulated values
+     * are set to 0.0.
+     */
+    public void clear() {
+        DoubleTools.fill(this.accu, 0, this.quanta, 0.0);
+    }
+    
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public String toString() {
+        return "[" + DoubleTools.asString(this.accu, 2) + "]";
+    }
+    
+    /**
+     * Instantiates a Histogram object for given values value range determined
+     * by a lower and an upper bound, with a given number a quanta (bins). 
+     * <br></br>
+     * @param lbd Lower bound of the accumulation interval.
+     * @param ubd Upper bound of the accumulation interval.
+     * @param quanta Number of quanta (bins).
+     */
+    public Histogram(final double lbd, final double ubd, final int quanta) {
+        //
+        this.lbd     = Math.min(lbd,  ubd);
+        this.ubd     = Math.max(lbd,  ubd);
+        this.range   = this.ubd - this.lbd;
+        this.irange  = (this.range > 0.0)?(1.0 / this.range):(0.0);
+        //
+        this.quanta  = Math.max(1, quanta);
+        this.dquanta = ((double)this.quanta) - Double.MIN_VALUE;
+        this.accu    = new double[this.quanta];
+    }
+    
+    /**
+     * Check, if at least one of the histogram values is positive (> 0.0).
+     * <br></br>
+     * @return True if one value is positive, false otherwise.
+     */
+    public boolean hasPositiveValues() {
+        for (int i = 0; i < this.quanta; i++) {
+            if (this.accu[i] > 0.0) return true;
+        }
+        return false;
+    }
+    
+    /**
+     * Check, if at least one of the histogram values is negative (< 0.0).
+     * <br></br>
+     * @return True if one value is negative, false otherwise.
+     */
+    public boolean hasNegativeValues() {
+        for (int i = 0; i < this.quanta; i++) {
+            if (this.accu[i] < 0.0) return true;
+        }
+        return false;
+    }
+    
+    /**
+     * This method produces a simple bar-chart representation of the
+     * inner accumulator as a white-space aligned String. The given
+     * height defines the number of text lines of which the highest bar
+     * in the histogram may consists of. Note, that the resulting
+     * string is just for ascii-style printouts, it should not be
+     * used for extracting histogram data.
+     * <br></br>
+     * @param height The number of lines of the highest bar.
+     * @return A string containing the bar-char representation of the histogram. 
+     */
+    public String printBars(final int height) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 18. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code blocks on lines 283-318, lines 325-358 into separate methods.